### PR TITLE
[Python3] cross build support

### PR DESCRIPTION
**Pull request by rinechran**
**Describe the pull request**

- #### What does your PR fix?
  Python3 cross support

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  all

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Your answer

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

### CODE REVIEW

**Comment by Osyotr** [comment]
Maybe you need to set `VCPKG_MAKE_BUILD_TRIPLET` in your custom triplet.
https://github.com/Microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_configure_make.md#build_triplet

**Comment by rinechran** [comment]
Are you saying that I should let you do VCPKG_MAKE_BUILD_TRIPLET when it's VCPKG_CROSS COMPILING?

**Comment by dg0yt** [comment]
The first step is to add `DETERMINE_BUILD_TRIPLET` to `vcpkg_cmake_configure`. `--host` is an argument which takes a value, and the proposed keyword will handle this for common setups. 
`VCPKG_MAKE_BUILD_TRIPLET` is a (vcpkg) triplet variable which can override the automatically determined value  for less common (or unsupported) setups.

Note that there can be more blockers for cross builds, in particular if the package tries to build some internal tools for the build maching.

**Comment by Osyotr** [comment]
@dg0yt would you kindly help me understand why does `BUILD_TRIPLET` argument exists if its value depends on toolchain and why `DETERMINE_BUILD_TRIPLET` is not default?

**Comment by dg0yt** [comment]
> @dg0yt would you kindly help me understand why does `BUILD_TRIPLET` argument exists if its value depends on toolchain and why `DETERMINE_BUILD_TRIPLET` is not default?

I could only speculate: The authors didn't want to rely on a particular automatic detection, or they didn't want to change the default behaviour when the keyword was added.
Note that `AUTOCONFIG` also implies `DETERMINE_BUILD_TRIPLET`.
And there are more quirks. There is no CI coverage for these kinds of cross builds.

**Comment by rinechran** [comment]
> > @dg0yt would you kindly help me understand why does `BUILD_TRIPLET` argument exists if its value depends on toolchain and why `DETERMINE_BUILD_TRIPLET` is not default?
> 
> I could only speculate: The authors didn't want to rely on a particular automatic detection, or they didn't want to change the default behaviour when the keyword was added. Note that `AUTOCONFIG` also implies `DETERMINE_BUILD_TRIPLET`. And there are more quirks. There is no CI coverage for these kinds of cross builds.

i respect your opinion .the, it seem right to add  ```DETERMINE_BUILD_TRIPLET ```

**Comment by Cheney-W** [comment]
Please modify the value of `"port-version"` from `4` to `5` in[ vcpkg.json](https://github.com/microsoft/vcpkg/blob/master/ports/python3/vcpkg.json) and run `.\vcpkg x-add-version python3 --overwrite-version`

**Comment by rinechran** [comment]
The "cpkg x-add-version" version does not go up because of the merge. What should I do?

File: ports/python3/portfile.cmake [hunk 1]
```diff
@@ -224,12 +224,18 @@ else()
         "--without-readline"
         "--disable-test-modules"
     )
+    if (VCPKG_CROSSCOMPILING)
+        list(APPEND OPTIONS "ac_cv_file__dev_ptmx=no")
```
Comment (rinechran) [inline ports/python3/portfile.cmake:228]: ,

**Comment by rinechran** [comment]
the fetch didn't work , but i don't thinkl there's a big problem with my code

**Comment by rinechran** [comment]
@Cheney-W  I think there's a ci bug.

**Comment by Cheney-W** [comment]
Waiting https://github.com/microsoft/vcpkg/pull/27973 merged.

**Comment by rinechran** [comment]
Can I close this pr?

**Comment by Cheney-W** [comment]
> Can I close this pr?

Sorry, I forgot it. 
No. The dependence PR has been merged, I will rerun this PR, if the CI test pass, we will consider merge this PR.

**Comment by rinechran** [comment]
please check the pr @Cheney-W

**Comment by autoantwort** [comment]
This change breaks the build of python:arm64-osx on an arm64 system

**Comment by rinechran** [comment]
It's a flag that only works when crossbuilding, so I don't think it'll affect when building locally on "python:arm64-osx". @autoantwort

**Comment by autoantwort** [comment]
But it does. Now building python fails with 
```
configure: WARNING: unrecognized options: --disable-silent-rules, --enable-static
configure: error: Cross compiling required --host=HOST-TUPLE and --build=ARCH
```
And before this change even cross building already works, now it also fails with the same error.

**Comment by rinechran** [comment]
Ha... 
I think this error may be caused by the DETERMINE_BUILD_TRIPLET flag

**Comment by autoantwort** [comment]
Do you have an example of a build failure that was fixed by this PR or do we only have examples of things that are now broken after this PR?

**Comment by rinechran** [comment]
> 

I'm not sure about that.

**Comment by autoantwort** [comment]
So what was the motivation behind creating this PR?

**Comment by rinechran** [comment]
An error occurs when cross-build python.
I added a version of the library by referring to the site below
https://bugs.python.org/issue3754 , https://unix.stackexchange.com/questions/342925/cross-compiling-python

**Comment by autoantwort** [comment]
Can you give an example from which to which triplet you wanted to cross compile where it failed?

**Comment by Osyotr** [comment]
`DETERMINE_BUILD_TRIPLET` shall be eradicated (maybe https://github.com/microsoft/vcpkg/pull/28374 could help).
`ac_cv_file__dev_ptmx` and `ac_cv_file__dev_ptc` values should come from your custom triplet.

**Comment by autoantwort** [comment]
Even with https://github.com/microsoft/vcpkg/pull/28374 I get the following error when I try to build python3:arm64-osx on my arm64-osx mac:
```
configure: WARNING: unrecognized options: --disable-silent-rules, --enable-static
configure: error: cross build not supported for aarch64-apple-darwin
```

**Comment by dpogue** [comment]
@autoantwort Ironically, I opened that PR specifically to try to work around issues cross-compiling arm64-osx Python from an x86_64 macOS builder, but alas (as you know) Python doesn't support cross-compiling on macOS: https://github.com/python/cpython/issues/90905 😞

**Pull request merged**